# programming-workshop

To run the code in the example files and examples in the workshop, you will need to install and download a couple of resources.

## Setting up your programming environment

If you do not have R and an integrated development environment (IDE) to develop R programs, you will need to install these first. Please follow [this guide](https://rstudio-education.github.io/hopr/starting.html) to install R and RStudio for your platform.

## Getting the workshop files from GitHub to your computer

The workshop is hosted on GitHub in [this repository](https://github.com/johanzvrskovec/programming-workshop). You can either clone this repository using Git, or download the content by 1.) clicking the large green button saying "\< \> Code", 2.) selecting "Download ZIP", 3.) unpack the downloaded ZIP file. It should contain the workshop project directory.

The workshop content is in `programming-workshop.Rmd`, located in the root of the project folder. This file may also contain these instructions repeated over again.

## Data files

To run the example analysis scripts you will need two data files. Both of these files are to be put into the project `data` folder.

### The reference panel file

One data file; `reference.1000G.maf.0.005.txt.gz`, contains a reference list of genetic variants. Please download the variant reference file from [here](https://utexas.app.box.com/s/vkd36n197m8klbaio3yzoxsee6sxo11v).

### The Genome-Wide Association Study summary statistics file

The other data file; `GCST90027164_buildGRCh37.tsv.gz`, contains summary statistics from a Genome-Wide Association Study (GWAS) performed on Amyotrophic lateral sclerosis (ALS). This file can be downloaded from [here](https://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/GCST90027001-GCST90028000/GCST90027164/).

Do not forget to put the downloaded files into the project `data` folder for the workshop scripts to work.

## 
